### ISSUE TITLE
Linkerd showed Service in fail-fast error while master node is down

### ISSUE THREAD

**Issue by DesmondH0** [linkerd/linkerd2#5582]
### What is the issue?
We occasionally found that while master node is down, the intra-node connection will be failing.
And we also found the linkerd-proxy is reporting `Service in fail-fast` error as below:
```
[178882.561866s]  WARN ThreadId(01) outbound:accept{peer.addr=10.52.1.11:51442 target.addr=10.118.5.159:8080}: linkerd2_app_core::errors: Failed to proxy request: Service in fail-fast
```

I expect while master node is gone, services should be still working and intra-node connection should be still able to connect.

### How can it be reproduced?
The event was found while GKE upgrading a Zonal Cluster.
I guess it would be reproducible while master node is down in any cases

### Logs, error output, etc

From linkerd-proxy injected to our application pods:
```
[178882.561866s]  WARN ThreadId(01) outbound:accept{peer.addr=10.52.1.11:51442 target.addr=10.118.5.159:8080}: linkerd2_app_core::errors: Failed to proxy request: Service in fail-fast
```

And there are nothing from controller while master node is gone, the histogram is like this for linkerd controller
![image](https://user-images.githubusercontent.com/39398132/105271012-b7826280-5bd1-11eb-9ce8-e2adf8015000.png)

#### `linkerd check` output

because we were tracing a historic events, didn't have a chance to type `linkerd check` at that moment.

### Environment

- Kubernetes Version: v1.17.14-gke.400
- Cluster Environment: GKE
- Host OS: Container-Optimized OS from Google
- Linkerd version: v2.9.1

### Possible solution
This is happening on GKE Zonal Cluster (single master)
I guess this may not be a problem with multi-master cluster

**Comment by olix0r** [comment]
@DesmondH0 When you say "master node", you're referring to the Kubernetes API server?

I assume if the API server was down, then no pods were rescheduled during that time.

Linkerd's load balancer should remain cached as long as traffic is running through the system, though it could idle out if traffic stopped for some time.

I'd be curious to check the linkerd-destination controller's logs during this time...

To summarize: we should continue to process traffic in many cases, but if new resolutions need to be performed, it may be impossible for us to perform discovery. We'd have to narrow down details of this situation to know whether this was a case where we could process requests normally or whether it would be impossible for us to process requests...

**Comment by DesmondH0** [comment]
Hi @olix0r, sorry for confusion.
Yes, i am referring to K8S API server.

And from the destination pod logs, it seems quite normal back at that moment following are some examples:
```
2021-01-20 04:04:08.918 HKT
time="2021-01-19T20:04:08Z" level=info msg="Establishing watch on service prod/some-prod-services" addr=":8086" component=traffic-split-watcher
2021-01-20 04:04:08.918 HKT
time="2021-01-19T20:04:08Z" level=info msg="Establishing watch on profile prod/some-prod-services.prod.svc.cluster.local" addr=":8086" component=profile-watcher
2021-01-20 04:04:08.918 HKT
time="2021-01-19T20:04:08Z" level=info msg="Establishing watch on profile prod/some-prod-services.prod.svc.cluster.local" addr=":8086" component=profile-watcher
2021-01-20 04:04:08.925 HKT
time="2021-01-19T20:04:08Z" level=info msg="Establishing watch on endpoint [prod/some-prod-services:8080]" addr=":8086" component=endpoints-watcher
2021-01-20 04:04:09.078 HKT
time="2021-01-19T20:04:09Z" level=info msg="Stopping watch on profile prod/some-prod-services.prod.svc.cluster.local" addr=":8086" component=profile-watcher
2021-01-20 04:04:09.078 HKT
time="2021-01-19T20:04:09Z" level=info msg="Stopping watch on profile prod/some-prod-services.prod.svc.cluster.local" addr=":8086" component=profile-watcher
2021-01-20 04:04:09.078 HKT
time="2021-01-19T20:04:09Z" level=info msg="Stopping watch on profile prod/some-prod-services" addr=":8086" component=traffic-split-watcher
```

And from the external alert system, the external traffic going into the cluster (public ingress) is returning 503 at that moment.
I guess its related to the linkerd proxy error i mentioned above 🤔

**Comment by adleong** [comment]
Hi @DesmondH0!  Unfortunately, we would need a lot more information to be able to look into this issue.  Detailed steps to reproduce it would be best.  In absence of that, more complete information about the state of your system would be helpful.  For example: from the proxy logs you posted, it appears that the proxy is attempting to send a request to `10.118.5.159:8080` but goes into fail-fast.  Is there a live pod at that address which is listening on that port?  Does the source pod have network connectivity to that address?

**Comment by adleong** [comment]
Closing due to inactivity.  Please feel free to re-open if you are able to provide more information.
